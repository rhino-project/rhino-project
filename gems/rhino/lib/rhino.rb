# frozen_string_literal: true

require 'rhino/engine'
require 'validators/country_validator'
require 'validators/email_validator'
require 'active_support'

module Rhino
  extend ActiveSupport::Autoload
  extend ActiveSupport::Concern

  autoload :SieveStack, 'rhino/sieve'

  # The root path for the api ie '/api'
  mattr_accessor :namespace, default: :api

  # Include Rhino::Resource::ActiveRecordExtension by default
  mattr_accessor :auto_include_active_record, default: true

  mattr_accessor :resources, default: if Rails.env.production?
                                        ['ActiveStorage::Attachment']
                                      else
                                        ['ActiveStorage::Attachment', 'Rhino::OpenApiInfo', 'Rhino::InfoGraph']
                                      end

  mattr_accessor :resource_classes

  mattr_accessor :registered_modules, default: {}

  # sieves
  mattr_accessor :sieves

  ##
  # Stolen from devise
  #
  # https://github.com/jinzhu/devise/blob/b94ee9da98b16e4c8fbdc91af8605669d01b17e6/lib/devise.rb#L239
  class Getter
    def initialize(name)
      @name = name
    end

    def get
      ActiveSupport::Dependencies.constantize(@name)
    end
  end

  def self.ref(arg)
    ActiveSupport::Dependencies.reference(arg)
    Getter.new(arg)
  end
  #
  ##

  # Get the resource classes from the resource reference object.
  def self.resource_classes
    resource_classes ||= resources.map(&:constantize)

    resource_classes
  end

  self.sieves = Rhino::SieveStack.new do |sieve|
    sieve.use Rhino::Sieve::Filter

    sieve.use Rhino::Sieve::Search

    sieve.use Rhino::Sieve::Order

    sieve.use Rhino::Sieve::Offset

    sieve.use Rhino::Sieve::Limit, default_limit: 20
  end

  def self.auth_owner
    @@auth_owner_ref.get
  end

  # Set the auth owner reference object to access the mailer.
  def self.auth_owner=(class_name)
    @@auth_owner_ref = ref(class_name) # rubocop:disable Style/ClassVars
  end
  self.auth_owner = 'User'

  def self.auth_owner_sym
    auth_owner.to_s.underscore.pluralize.to_sym
  end

  def self.base_owner
    @@base_owner_ref.get
  end

  # Set the mailer reference object to access the mailer.
  def self.base_owner=(class_name)
    @@base_owner_ref = ref(class_name) # rubocop:disable Style/ClassVars
  end
  self.base_owner = 'User'

  def self.base_owner_sym
    base_owner.to_s.underscore.pluralize.to_sym
  end

  def self.same_owner?
    base_owner == auth_owner
  end

  def self.base_to_auth
    return auth_owner.model_name.i18n_key if same_owner?

    return auth_owner_sym if base_owner.reflections.key?(auth_owner_sym.to_s)

    nil
  end

  def self.auth_to_base
    return auth_owner.model_name.i18n_key if same_owner?

    return base_owner_sym if auth_owner.reflections.key?(base_owner_sym.to_s)

    nil
  end

  # Default way to set up Rhino
  def self.setup
    yield self
  end
end
