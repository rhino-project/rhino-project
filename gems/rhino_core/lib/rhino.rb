# frozen_string_literal: true

require 'rhino/engine'
require 'active_support'

module Rhino
  extend ActiveSupport::Concern

  # The root path for the api ie '/api'
  mattr_accessor :namespace, default: :api

  # Base owner
  mattr_accessor :base_owner, default: :user

  # List of resources
  mattr_accessor :resources, default: []

  ##
  # Stolen from devise
  #
  # https://github.com/jinzhu/devise/blob/b94ee9da98b16e4c8fbdc91af8605669d01b17e6/lib/devise.rb#L239
  class Getter
    def initialize(name)
      @name = name
    end

    def get
      ActiveSupport::Dependencies.constantize(@name)
    end
  end

  def self.ref(arg)
    ActiveSupport::Dependencies.reference(arg)
    Getter.new(arg)
  end
  #
  ##

  # Get the resource classes from the resource reference object.
  def self.resources
    @@resource_refs.map(&:get)
  end

  # Set the resource classes reference array to access the resource classes.
  def self.resources=(class_names)
    @@resource_refs = class_names.map { |class_name| ref(class_name) } # rubocop:disable Style/ClassVars
  end
  self.resources = []

  # List of models that need to be wrapped in MetaModel
  mattr_accessor :proxy_models, default: {}

  # FIXME: Cache this - Devise has some nice code in devise.rb to do this
  def self.base_owner_class
    base_owner.to_s.humanize.classify.constantize
  end

  def self.resource_from_path(path)
    path.remove(%r{^\/#{namespace}\/}).split('/').first
  end

  def self.resource_by_association(association)
    proxy_models[association.options[:class_name]]&.safe_constantize || association.klass
  end

  def self.resource_instance(inst)
    return proxy_models[inst.class.name].constantize.new inst if proxy_models.key?(inst.class.name)

    inst
  end

  # Default way to set up Rhino
  def self.setup
    yield self
  end
end
