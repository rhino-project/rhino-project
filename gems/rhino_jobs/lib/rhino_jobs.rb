# frozen_string_literal: true

require 'rhino_jobs/engine'

module RhinoJobs
  # Your code goes here...
end
