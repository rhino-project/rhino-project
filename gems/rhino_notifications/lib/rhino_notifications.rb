# frozen_string_literal: true

require "rhino_notifications/engine"

module RhinoNotifications
  # Your code goes here...
end
