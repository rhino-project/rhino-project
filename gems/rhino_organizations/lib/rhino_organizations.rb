# frozen_string_literal: true

require 'rhino_organizations/engine'

module RhinoOrganizations
end
