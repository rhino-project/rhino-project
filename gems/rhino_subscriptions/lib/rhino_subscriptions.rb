# frozen_string_literal: true

require 'rhino_subscriptions/engine'
# require 'active_support'

module RhinoSubscriptions
  # Your code goes here...
end
