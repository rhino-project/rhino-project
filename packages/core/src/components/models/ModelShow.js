import PropTypes from 'prop-types';
import { useGlobalComponent, useOverrides } from 'rhino/hooks/overrides';
import ModelShowDescription from 'rhino/components/models/ModelShowDescription';
import ModelShowRelated from 'rhino/components/models/ModelShowRelated';
import ModelShowActions from 'rhino/components/models/ModelShowActions';
import ModelShowSimple from './ModelShowSimple';
import ModelShowHeader from './ModelShowHeader';
import ModelSection from './ModelSection';

const defaultComponents = {
  ModelShowHeader,
  ModelShowActions,
  ModelShowDescription,
  ModelShowRelated
};

export const ModelShowBase = ({ overrides, ...props }) => {
  const {
    ModelShowHeader,
    ModelShowActions,
    ModelShowDescription,
    ModelShowRelated
  } = useOverrides(defaultComponents, overrides);

  if (ModelShowDescription().props?.paths)
    console.warn('ModelShowDescription pass legacy paths prop');

  return (
    <ModelShowSimple paths={ModelShowDescription().props?.paths} {...props}>
      <ModelSection baseClassName="show">
        <ModelShowHeader />
        <ModelShowActions />
        <ModelShowDescription />
        <ModelShowRelated />
      </ModelSection>
    </ModelShowSimple>
  );
};

ModelShowBase.propTypes = {
  model: PropTypes.oneOfType([PropTypes.string, PropTypes.object]).isRequired,
  modelId: PropTypes.string.isRequired,
  overrides: PropTypes.object
};

const ModelShow = (props) => useGlobalComponent(ModelShowBase, props);

export default ModelShow;
