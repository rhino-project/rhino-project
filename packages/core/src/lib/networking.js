import axios from 'axios';
import qs from 'qs';
import * as networking from './networking.js';

import env from 'config';
import { toastStore } from 'rhino/queries/toast';

const REACT_APP_API_ROOT_PATH = env.REACT_APP_API_ROOT_PATH;
export const AUTH_BASE_PATH = 'api/auth';
export const AUTH_ACCEPT_PATH = `${AUTH_BASE_PATH}/invitation`;
export const AUTH_CREATE_END_POINT = AUTH_BASE_PATH + '/sign_in';
export const AUTH_DESTROY_END_POINT = AUTH_BASE_PATH + '/sign_out';
export const AUTH_PASSWORD_END_POINT = AUTH_BASE_PATH + '/password';
export const AUTH_VALIDATE_TOKEN_END_POINT = AUTH_BASE_PATH + '/validate_token';
export const AUTH_SESSION_KEY = ['session'];

const DEFAULT_HEADERS = {
  'Content-Type': 'application/json',
  Accept: 'application/json'
};

export const constructPath = (path) =>
  new URL(path, REACT_APP_API_ROOT_PATH).toString();

const _buildHeaders = (headers = {}) => {
  return { ...DEFAULT_HEADERS, ...headers };
};

export class NetworkUnauthorizedError extends Error {
  constructor(...params) {
    // Pass remaining arguments (including vendor specific ones) to parent constructor
    super(...params);

    this.name = 'NetworkUnauthorizedError';
  }
}

export class NetworkParamError extends Error {
  constructor(errors, ...params) {
    // Pass remaining arguments (including vendor specific ones) to parent constructor
    super(...params);

    this.name = 'NetworkParamError';
    this.errors = errors;
  }
}

export const networkApiCall = (path, options) => {
  const defaultOptions = {
    method: 'get',
    headers: {},
    data: null,
    signal: null,
    ...options
  };

  return axios(constructPath(path), {
    ...defaultOptions,
    headers: _buildHeaders(defaultOptions.headers),
    paramsSerializer: {
      serialize: (params) => qs.stringify(params, { arrayFormat: 'brackets' })
    },
    withCredentials: true
  }).catch((error) => {
    if (
      ((error.response.status === 401 || error.response.status === 403) &&
        (!path.startsWith('api/auth') || path === 'api/auth/validate_token')) ||
      (error.response.status === 404 && path === 'api/auth/sign_out')
    ) {
      // If the response is 401 or 403 (and not part of authenticating (other
      // than token validation), invalidate the session
      throw new NetworkUnauthorizedError();
    } else if (error.response.status === 500) {
      // If its a 500 throw up a toast
      // FIXME De-duplicate these
      toastStore.add({
        icon: 'danger',
        title: 'Request Error',
        description: 'Sorry, the server returned an error during your request'
      });
    } else if (error.response.data?.errors) {
      // If its a "regular" error, pass the messages back to the user
      throw new NetworkParamError(error.response.data?.errors);
    }
  });
};

const handler = {
  get(target, prop, receiver) {
    // If the data is being access data.data, its the older form, return it
    if (prop === 'data') {
      console.warn('Legacy data access used in query hooks');

      return target;
    }
    return Reflect.get(...arguments);
  }
};

export const networkApiCallOnlyData = async (
  path,
  options = { method: 'get', headers: {}, data: null, signal: null }
) => {
  const response = await networking.networkApiCall(path, options);

  return new Proxy(response.data, handler);
};
