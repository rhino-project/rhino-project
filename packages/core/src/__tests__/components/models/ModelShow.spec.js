import ModelShow from 'rhino/components/models/ModelShow';
import { sharedModelTests } from './sharedModelTests';
import { render } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import * as network from 'rhino/lib/networking';
import rhinoConfig from 'rhino.config';

vi.spyOn(network, 'networkApiCall').mockReturnValue({
  data: {
    test: 'test'
  }
});

describe('ModelShow', () => {
  const Foo = () => <div>Foo</div>;
  const Bar = () => <div>Bar</div>;
  const Baz = () => <div>Baz</div>;
  const Foz = () => <div>Foz</div>;

  const wrapper = ({ children }) => {
    const queryClient = new QueryClient({
      defaultOptions: {
        queries: {
          retry: false
        }
      }
    });

    return (
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    );
  };

  sharedModelTests(ModelShow);

  describe('local overrides', () => {
    let configSpy;

    beforeEach(() => {
      configSpy = vi.spyOn(rhinoConfig, 'components', 'get').mockReturnValue({
        ModelShowHeader: null,
        ModelShowActions: {
          props: {
            paths: ['title', 'category']
          }
        },
        ModelShowDescription: null,
        ModelShowRelated: null
      });
    });

    afterEach(() => {
      configSpy.mockRestore();
    });

    it(`should allow local overrides`, () => {
      const { asFragment } = render(
        <ModelShow
          overrides={{
            ModelShowHeader: Foo,
            ModelShowActions: Baz,
            ModelShowDescription: Bar,
            ModelShowRelated: Foz
          }}
          model={{ model: 'user', properties: { name: {} } }}
          modelId="1"
          path="name"
          fallback={false}
        />,
        { wrapper }
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it(`should use global overrides when local are not present`, () => {
      const { asFragment } = render(
        <ModelShow
          model={{ model: 'user', properties: { name: {} } }}
          modelId="1"
          path="name"
          fallback={false}
        />,
        { wrapper }
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
