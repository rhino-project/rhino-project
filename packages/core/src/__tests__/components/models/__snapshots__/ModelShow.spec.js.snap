// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ModelShow > should allow local overrides 1`] = `
<DocumentFragment>
  <div
    class="model-show model-show-user"
  >
    <div>
      Foo
    </div>
    <div>
      Baz
    </div>
    <div>
      Bar
    </div>
    <div>
      Foz
    </div>
  </div>
</DocumentFragment>
`;

exports[`ModelShow > should render with global override shorthand 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;

exports[`ModelShow > should render with global override shorthand for model 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;
